src/engine/Server.cpp

```diff
@@ -492,6 +494,139 @@ Awaitable<void> Server::process(
+  if (!std::holds_alternative<updateClause::GraphUpdate>(updateClause.op_)) {
+    throw std::runtime_error(
+        "Only INSERT/DELETE update operations are currently supported.");
+  }
+  auto graphUpdate = std::get<updateClause::GraphUpdate>(updateClause.op_);
+  auto res = qet.getResult(true);
+
+  auto& vocab = index_.getVocab();
+  // DeltaTriples transfers the IDs that are actually inserted to its own
+  // LocalVocab. This LocalVocab only contains IDs that are related to the
+  // template. Most of the IDs will be added to the DeltaTriples' LocalVocab. An
+  // ID will not be added if it belongs to a Quad with a variable that has no
+  // solutions.
+  LocalVocab localVocab = {};
+  // The GraphUpdate consist of a template of `SparqlTripleSimpleWithGraph` and
+  // in some cases also an GraphPattern. The template contains
+  // `TripleComponent`s (s, p and o) and a `Graph` (graph). First transform
+  // everything except Variables in the template to IDs.
+  using IdOrVariable = std::variant<Id, Variable>;
+  auto transformSparqlTripleComponent =
+      [&vocab, &localVocab](TripleComponent component) -> IdOrVariable {
+    if (component.isVariable()) {
+      return std::move(component.getVariable());
+    } else {
+      return std::move(component).toValueId(vocab, localVocab);
+    }
+  };
+  auto transformGraph =
+      [&vocab,
+       &localVocab](SparqlTripleSimpleWithGraph::Graph graph) -> IdOrVariable {
+    return std::visit(
+        ad_utility::OverloadCallOperator{
+            [](const std::monostate&) -> IdOrVariable {
+              AD_CORRECTNESS_CHECK(
+                  qlever::specialIds().contains(DEFAULT_GRAPH_IRI));
+              return qlever::specialIds().at(DEFAULT_GRAPH_IRI);
+            },
+            [&vocab, &localVocab](const Iri& iri) -> IdOrVariable {
+              ad_utility::triple_component::Iri i =
+                  ad_utility::triple_component::Iri::fromIriref(iri.iri());
+              return TripleComponent(i).toValueId(vocab, localVocab);
+            },
+            [](const Variable& var) -> IdOrVariable { return var; }},
+        graph);
+  };
+  auto transformSparqlTripleSimple =
+      [&transformSparqlTripleComponent,
+       &transformGraph](SparqlTripleSimpleWithGraph triple) {
+        return std::array{transformSparqlTripleComponent(std::move(triple.s_)),
+                          transformSparqlTripleComponent(std::move(triple.p_)),
+                          transformSparqlTripleComponent(std::move(triple.o_)),
+                          transformGraph(std::move(triple.g_))};
+      };
+  std::vector<std::array<IdOrVariable, 4>> toInsertTemplates =
+      ad_utility::transform(std::move(graphUpdate.toInsert_),
+                            transformSparqlTripleSimple);
+  std::vector<std::array<IdOrVariable, 4>> toDeleteTemplates =
+      ad_utility::transform(std::move(graphUpdate.toDelete_),
+                            transformSparqlTripleSimple);
```

Comment:
All this transformation 
can be In a separate Function (for testability and less file length) and also in a separate .cpp file
(Transform the Update-Template, and return the Template + local Vocab is a nice encapsulation.